File: packages/lite-filter/npm/src/ComparisonOperator.ts
/**
 * ComparisonOperator class - represents all supported comparison operators
 * This is an internal implementation detail
 */
export class ComparisonOperator {
  /**
   * The token as it appears in filter syntax (e.g., "=", ":contains:", ">=")
   */
  readonly token: string;

  /**
   * The programmatic name (e.g., "EQUALS", "CONTAINS", "GREATER_THAN_OR_EQUAL")
   */
  readonly name: string;

  private constructor(token: string, name: string) {
    this.token = token;
    this.name = name;
  }

  // Standard RFC4515 operators (no colons)
  static readonly EQUALS = new ComparisonOperator('=', 'EQUALS');
  static readonly NOT_EQUALS = new ComparisonOperator('!=', 'NOT_EQUALS');
  static readonly GREATER_THAN = new ComparisonOperator('>', 'GREATER_THAN');
  static readonly GREATER_THAN_OR_EQUAL = new ComparisonOperator('>=', 'GREATER_THAN_OR_EQUAL');
  static readonly LESS_THAN = new ComparisonOperator('<', 'LESS_THAN');
  static readonly LESS_THAN_OR_EQUAL = new ComparisonOperator('<=', 'LESS_THAN_OR_EQUAL');
  static readonly APPROX_MATCH = new ComparisonOperator('~=', 'APPROX_MATCH');
  static readonly PRESENT = new ComparisonOperator('=*', 'PRESENT');

  // Custom extensions (with colons)
  static readonly CONTAINS = new ComparisonOperator(':contains:', 'CONTAINS');
  static readonly STARTS_WITH = new ComparisonOperator(':startsWith:', 'STARTS_WITH');
  static readonly ENDS_WITH = new ComparisonOperator(':endsWith:', 'ENDS_WITH');
  static readonly MATCHES = new ComparisonOperator(':matches:', 'MATCHES');
  static readonly IS_NULL = new ComparisonOperator(':isnull:', 'IS_NULL');
  static readonly IS_EMPTY = new ComparisonOperator(':isempty:', 'IS_EMPTY');
  static readonly INCLUDES = new ComparisonOperator(':includes:', 'INCLUDES');
  static readonly INCLUDES_ANY = new ComparisonOperator(':includesAny:', 'INCLUDES_ANY');
  static readonly BETWEEN = new ComparisonOperator(':between:', 'BETWEEN');
  static readonly WITHIN_DAYS = new ComparisonOperator(':withinDays:', 'WITHIN_DAYS');
  static readonly YEAR = new ComparisonOperator(':year:', 'YEAR');

  /**
   * All registered operators
   */
  private static readonly ALL_OPERATORS = [
    ComparisonOperator.EQUALS,
    ComparisonOperator.NOT_EQUALS,
    ComparisonOperator.GREATER_THAN,
    ComparisonOperator.GREATER_THAN_OR_EQUAL,
    ComparisonOperator.LESS_THAN,
    ComparisonOperator.LESS_THAN_OR_EQUAL,
    ComparisonOperator.APPROX_MATCH,
    ComparisonOperator.PRESENT,
    ComparisonOperator.CONTAINS,
    ComparisonOperator.STARTS_WITH,
    ComparisonOperator.ENDS_WITH,
    ComparisonOperator.MATCHES,
    ComparisonOperator.IS_NULL,
    ComparisonOperator.IS_EMPTY,
    ComparisonOperator.INCLUDES,
    ComparisonOperator.INCLUDES_ANY,
    ComparisonOperator.BETWEEN,
    ComparisonOperator.WITHIN_DAYS,
    ComparisonOperator.YEAR,
  ];

  /**
   * Parses a token string and returns the corresponding ComparisonOperator
   * @param token The token to parse (e.g., "=", ":contains:")
   * @returns The matching ComparisonOperator
   * @throws Error if token is not recognized
   */
  static parse(token: string): ComparisonOperator {
    const operator = ComparisonOperator.ALL_OPERATORS.find(op => op.token === token);
    if (!operator) {
      throw new Error(`Unknown comparison operator: ${token}`);
    }
    return operator;
  }
}

File: packages/lite-filter/java/src/main/java/com/zerobias/litefilter/ComparisonOperator.java
package com.zerobias.litefilter;

/**
 * Comparison operators for filter expressions.
 */
public enum ComparisonOperator {
    // Standard RFC4515 operators
    EQUALS("=", "EQUALS"),
    NOT_EQUALS("!=", "NOT_EQUALS"),
    GREATER_THAN(">", "GREATER_THAN"),
    GREATER_THAN_OR_EQUAL(">=", "GREATER_THAN_OR_EQUAL"),
    LESS_THAN("<", "LESS_THAN"),
    LESS_THAN_OR_EQUAL("<=", "LESS_THAN_OR_EQUAL"),
    APPROX_MATCH("~=", "APPROX_MATCH"),
    PRESENCE_CHECK("=*", "PRESENCE_CHECK"),

    // String operations (custom extensions with colons)
    CONTAINS(":contains:", "CONTAINS"),
    BEGINS_WITH(":startsWith:", "BEGINS_WITH"),
    ENDS_WITH(":endsWith:", "ENDS_WITH"),
    REGEX(":matches:", "REGEX"),

    // Null/empty checks (custom extensions)
    IS_NULL(":isnull:", "IS_NULL"),
    IS_EMPTY(":isempty:", "IS_EMPTY"),

    // Array operations (custom extensions)
    INCLUDES(":includes:", "INCLUDES"),
    INCLUDES_ANY(":includesAny:", "INCLUDES_ANY"),

    // Numeric operations (custom extensions)
    BETWEEN(":between:", "BETWEEN"),

    // Date operations (custom extensions)
    WITHIN_DAYS(":withinDays:", "WITHIN_DAYS"),
    YEAR(":year:", "YEAR");

    private final String token;
    private final String name;

    ComparisonOperator(String token, String name) {
        this.token = token;
        this.name = name;
    }

    /**
     * Get the token representation of this operator.
     *
     * @return the token string
     */
    public String getToken() {
        return token;
    }

    /**
     * Get the name of this operator.
     *
     * @return the name string
     */
    public String getName() {
        return name;
    }

    /**
     * Parse a token string to get the corresponding operator.
     *
     * @param token the token to parse
     * @return the comparison operator
     * @throws IllegalArgumentException if token is not recognized
     */
    public static ComparisonOperator parse(String token) {
        for (ComparisonOperator op : values()) {
            if (op.token.equals(token)) {
                return op;
            }
        }
        throw new IllegalArgumentException("Unknown operator token: " + token);
    }

    /**
     * Check if this operator is a custom extension (uses colon format).
     *
     * @return true if custom extension, false if standard operator
     */
    public boolean isCustomExtension() {
        return token.contains(":");
    }

    @Override
    public String toString() {
        return token;
    }
}
